### DISCUSSION TITLE
or-tools in another thread gives strange solution java

### DISCUSSION THREAD

**Discussion by nane-petrosyan** [google/or-tools#2744] [Routing (and legacy CP) questions]
Hello. I am working on route optimization using google or-tools with java. Everything works fine until I change the thread. I want to run `routing.solveWithParameters(searchParameters)` in another thread so that I can get progress with `routing.solver().topProgressPercent()` in my thread. This causes issues when I run my tests in parallel and I get different errors every time. I don't know what to do to make this process thread safe. Thank you in advance :)

**Comment by Mizux** [comment]
hi,
Currently the callback passed to the routing solver do not support thread change.

I'll be happy if could provide us a small example/snippet java code where you perform this thread change !
(or even better a small test to reproduce it ;))

On the same topic, I was working on a potential similar issue in the SAT solver this morning and just fix it !
See: https://github.com/google/or-tools/commit/e5a5b8e18fdfa3231662be5b54896763ac5689b0

So you may try to build from source a new java package and do the same in the following locations:
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/java/constraint_solver.i#L151-L155
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/java/constraint_solver.i#L178-L182
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/java/constraint_solver.i#L206-L215
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/java/constraint_solver.i#L239-L253

note: On my way to apply this "Fix" for routing solver on master (just need to find a way to reproduce it first)

**Reply by nane-petrosyan** [reply]
@Mizux thank you for your reply :) Here are some parts of my code.

```
new Thread(() -> {
            this.dataModel = dataModel;
            this.manager = createManager();

            this.routing = new RoutingModel(manager);

            addParamsToRoutingModel();

            RoutingSearchParameters searchParameters =
                    main.defaultRoutingSearchParameters()
                            .toBuilder()
                            .setSolutionLimit(dataModel.getMaxIterations())
                            .setTimeLimit(Duration.newBuilder().setSeconds(dataModel.getMaxTime()).build())
                            .setFirstSolutionStrategy(FirstSolutionStrategy.Value.forNumber(dataModel.getAlgorithmType()))
                            .setLocalSearchMetaheuristic(LocalSearchMetaheuristic.Value.forNumber(dataModel.getLocalAlgorithmType()))
                            .build();

             assignment = routing.solveWithParameters(searchParameters);
        }).start();
```

and here is the method for getting solution progress 

```
@Override
    public SolutionState getSolutionState() {
        waitUntilSetupFinished();
        solutionState.setPercentage(Math.max(routing.solver().topProgressPercent(), 0));

        return solutionState;
    }

    private void waitUntilSetupFinished() {
        while (routing == null) Thread.onSpinWait();
    }
```

p.s. SolutionState is an object created by me
